actors/builtin/miner/miner_actor.go

```diff
@@ -809,6 +810,114 @@ func (a Actor) PreCommitSector(rt Runtime, params *PreCommitSectorParams) *abi.E
+	AggregateProof []byte
+}
+
+const MaxAggregatedSectors = 819
+const MinAggregatedSectors = 1
+const MaxAggregateProofSize = 192000
+
+func (a Actor) ProveCommitAggregate(rt Runtime, params *ProveCommitAggregateParams) *abi.EmptyValue {
+	rt.ValidateImmediateCallerAcceptAny()
+	checkClaimExists(rt)
+	aggSectorsCount, err := params.SectorNumbers.Count()
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to count aggregated sectors")
+	if aggSectorsCount > MaxAggregatedSectors {
+		rt.Abortf(exitcode.ErrIllegalArgument, "too many sectors addressed")
+	} else if aggSectorsCount < MinAggregatedSectors {
+		rt.Abortf(exitcode.ErrIllegalArgument, "too few sectors addressed")
+	}
+
+	err = params.SectorNumbers.ForEach(func(sectorNo uint64) error {
+		if sectorNo > abi.MaxSectorNumber {
+			rt.Abortf(exitcode.ErrIllegalArgument, "sector number greater than maximum")
+		}
+		return nil
+	})
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check all sector numbers below max")
+
+	if uint64(len(params.AggregateProof)) > MaxAggregateProofSize {
+		rt.Abortf(exitcode.ErrIllegalArgument, "sector prove-commit proof of size %d exceeds max size of %d",
+			len(params.AggregateProof), MaxAggregateProofSize)
+	}
+
+	store := adt.AsStore(rt)
+	var st State
+	rt.StateReadonly(&st)
+
+	precommits, err := st.GetAllPrecommittedSectors(store, params.SectorNumbers)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to get precommits")
+
+	// compute data commitments
+	computeDataCommitmentsInputs := make([]*market.SectorDataCommitmentInputs, len(precommits))
+	for i, precommit := range precommits {
+		computeDataCommitmentsInputs[i] = &market.SectorDataCommitmentInputs{
+			SectorType: precommit.Info.SealProof,
+			DealIDs:    precommit.Info.DealIDs,
+		}
+	}
+	commDs := requestUnsealedSectorCIDs(rt, computeDataCommitmentsInputs...)
+
+	builtin.RequireState(rt, len(precommits) > 0, "bitfield non-empty but zero precommits read from state")
+	sealProof := precommits[0].Info.SealProof
+
+	svis := make([]proof.AggregateSealVerifyInfo, 0)
+
+	receiver := rt.Receiver()
+	minerActorID, err := addr.IDFromAddress(receiver)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "runtime provided non-ID receiver address %s", receiver)
+	buf := new(bytes.Buffer)
+	err = receiver.MarshalCBOR(buf)
+	receiverBytes := buf.Bytes()
+	builtin.RequireNoErr(rt, err, exitcode.ErrSerialization, "failed to marshal address for seal verification challenge")
+
+	for i, precommit := range precommits {
+		msd, ok := MaxProveCommitDuration[precommit.Info.SealProof]
+		if !ok {
+			rt.Abortf(exitcode.ErrIllegalState, "no max seal duration for proof type: %d", precommit.Info.SealProof)
+		}
+		proveCommitDue := precommit.PreCommitEpoch + msd
+		if rt.CurrEpoch() > proveCommitDue {
+			rt.Abortf(exitcode.ErrIllegalArgument, "commitment proof for %d too late at %d, due %d", precommit.Info.SectorNumber, rt.CurrEpoch(), proveCommitDue)
```

Comment:
> I assume we need the precommit information for all sectors in order to validate the proof. Could we note the sector as late, validate the proof, then drop the late sectors before the call to confirmSectorProofsValid?

I am unsure whether it's worth it for only between 0 and 30 minutes of buffer only after the first precommits have expired.  But I think this will be a clean change and therefore worth it.

Reply:
Yes, I think we should adjust the precommit cleanup too, either by doing it less often (e.g. daily, on deadline index == 0) or subtracting a day or so from the epoch that is considered expired, or both.

Reply:
I'm going to leave this whole behavior change to a follow-up PR